pg_search/src/postgres/transaction.rs

```diff
@@ -137,18 +137,57 @@ unsafe extern "C" fn pg_search_xact_callback(
 }
 
 fn finalize_drop<W: WriterClient<WriterRequest> + Send + Sync + 'static>(
-    writer: &Arc<Mutex<W>>,
+    _writer: &Arc<Mutex<W>>,
     directory: &WriterDirectory,
 ) -> Result<(), ClientError> {
-    // ask the remote side to drop the index
-    writer.lock().request(WriterRequest::DropIndex {
-        directory: directory.clone(),
-    })?;
+    let writer = unsafe { SearchIndex::get_writer() };
 
-    // then tell it to do it now
-    writer.lock().request(WriterRequest::Commit {
-        directory: directory.clone(),
-    })?;
+    writer
+        .drop_index(directory.clone())
+        .expect("must be able to finalize drop");
+    writer
+        .commit(directory.clone())
+        .expect("commit must work in finalize drop");
 
     Ok(())
 }
+
+/// Create a new index writer advisory lock.
+/// Advisory locks are global across the entire Postgres cluster,
+/// so its important that the key is qualified with the database oid.
+pub struct IndexWriterLock {
+    database_oid: u32,
+    index_oid: u32,
+}
+
+impl IndexWriterLock {
+    pub fn new(database_oid: u32, index_oid: u32) -> Self {
+        Self {
+            database_oid,
+            index_oid,
+        }
+    }
+
+    // This method combines the two OIDs into a single 64-bit lock key
+    fn key(&self) -> i64 {
+        // Shift the database_oid into the high 32 bits, then combine with index_oid
+        ((self.database_oid as i64) << 32) | (self.index_oid as i64)
+    }
+
+    pub fn acquire<T>(&self, func: impl FnOnce() -> T) -> T {
+        let lock_key = self.key(); // Get the lock key
+
+        // Attempt to acquire an exclusive lock on the generated key
+        // The lock will automatically be released at the end of the
+        // transaction.
+        unsafe {
+            direct_function_call::<()>(
+                pg_sys::pg_advisory_xact_lock_int8,
+                &[lock_key.into_datum()],
+            );
+        }
```

Comment:
Not suggesting we change anything yet, but I wonder if we could use the non-xact version?  Something to think about later.